pkg/vm/vm_test.go

```diff
@@ -3331,6 +3331,110 @@ func TestBitAndNumericOpcodes(t *testing.T) {
 	}
 }
 
+var stackIsolationTestCases = map[opcode.Opcode][]struct {
+	name   string
+	params []interface{}
+}{
+	opcode.CALLE: {
+		{
+			name: "no_params",
+		},
+		{
+			name:   "with_params",
+			params: []interface{}{big.NewInt(5), true, []byte{1, 2, 3}},
+		},
+	},
+	opcode.CALLED: {
+		{
+			name: "no_paranms",
+		},
+		{
+			name:   "with_params",
+			params: []interface{}{big.NewInt(5), true, []byte{1, 2, 3}},
+		},
+	},
+	opcode.CALLET: {
+		{
+			name: "no_params",
+		},
+		{
+			name:   "with_params",
+			params: []interface{}{big.NewInt(5), true, []byte{1, 2, 3}},
+		},
+	},
+	opcode.CALLEDT: {
+		{
+			name: "no_params",
+		},
+		{
+			name:   "with_params",
+			params: []interface{}{big.NewInt(5), true, []byte{1, 2, 3}},
+		},
+	},
+}
+
+func TestStackIsolationOpcodes(t *testing.T) {
+	scriptHash := util.Uint160{1, 2, 3}
+	for code, codeTestCases := range stackIsolationTestCases {
+		t.Run(code.String(), func(t *testing.T) {
+			for _, testCase := range codeTestCases {
+				t.Run(testCase.name, func(t *testing.T) {
+					rvcount := len(testCase.params) + 1 // parameters + DEPTH
+					pcount := len(testCase.params)
+					prog := []byte{byte(code), byte(rvcount), byte(pcount)}
+					if code == opcode.CALLE || code == opcode.CALLET {
+						prog = append(prog, scriptHash.BytesBE()...)
+					}
+					prog = append(prog, byte(opcode.RET))
+
+					vm := load(prog)
+					vm.SetScriptGetter(func(in util.Uint160) ([]byte, bool) {
+						if in.Equals(scriptHash) {
+							return makeProgram(opcode.DEPTH), true
+						}
+						return nil, false
+					})
+					if code == opcode.CALLED || code == opcode.CALLEDT {
+						vm.Context().hasDynamicInvoke = true
+					}
+					if code == opcode.CALLET || code == opcode.CALLEDT {
+						vm.Context().rvcount = rvcount
+					}
+					// add extra parameter just to check that it won't appear on new estack
+					vm.estack.PushVal(7)
+					for _, param := range testCase.params {
+						vm.estack.PushVal(param)
+					}
+					if code == opcode.CALLED || code == opcode.CALLEDT {
+						vm.estack.PushVal(scriptHash.BytesBE())
+					}
+
+					runVM(t, vm)
+					elem := vm.estack.Pop() // depth should be equal to params count, as it's a separate execution context
+					assert.Equal(t, int64(pcount), elem.BigInt().Int64())
+					for i := pcount; i > 0; i-- {
+						p := vm.estack.Pop()
+						assert.Equal(t, testCase.params[i-1], p.value.Value())
+					}
+				})
+			}
+		})
+	}
+}
+
+func TestCALLI(t *testing.T) {
+	prog := []byte{byte(opcode.PUSHBYTES2), byte(opcode.Opcode(uint16(1))), byte(opcode.Opcode(uint16(1) >> 8)), // little-endian
+		byte(opcode.TOALTSTACK), byte(opcode.DUPFROMALTSTACK),
+		byte(opcode.JMPIF), byte(0x4), byte(0), byte(opcode.RET),
+		byte(opcode.FROMALTSTACK), byte(opcode.DEC),
+		byte(opcode.CALLI), byte(0), byte(1), 0xF8, 0xFF} // -8 -> JMP to TOALTSTACK)
```

Comment:
This won't jump to `TOALTSTACK`, rather it will jump to `JMPIF`.

And I'd suggest something like this to test RV and stack isolation properly.

1) Check that alt stack is clean (it must panic on `FROMALTSTACK`):
```
PUSH1
DUP
TOALTSTACK
JMP 0x2 /* to CALLI */
FROMALTSTACK
RET
CALLI {0, 0, -2} /* to FROMALTSTACK */
RET
```

2) Check parameter and return values copying:
```
PUSH3
PUSH2
PUSH1
JMP 0x2 /* to CALLI */
SUB /* 2 - 1 */
RET
CALLI {1, 2, -4}  /* to SUB */
MUL /* 3 * 1 */
RET
```
It should run to completion and leave `3` on estack. Failing to copy `pcount` elements it would fail on `SUB`, doing it in wrong order it would compute `-1` instead on 1, not returning `rvcount` elements it would  fail on `MUL`, so it seems to be a nice test to me.

Estack isolation can also be checked with `DEPTH` as with `CALLE` tests.

Reply:
I finally got how it works!

Reply:
Added also DEPTH check